### DISCUSSION TITLE
ActivityPub Compatibility (Even if Revolt is Centralized)

### DISCUSSION THREAD

**Discussion by goliv04053** [stoatchat/discussions#469] [Feature Suggestions]
Even if Revolt is Centralized, why Revolt can't go into federation?

I understand that is necessary to analyze how it will be implemented. I think that it can be discussed

**Comment by insertish** [comment]
(note: i wrote this originally thinking about just federation, if we were to do ActivityPub we'd have to be a microblogging platform or some variant of one)

I mean, if you can plan it out and figure out a way to implement it, and then find someone to implement it, then sure.
The main problem is that I'm pretty much the only one working on stuff most of the time, and I already have a huge backlog, so the main thing is reducing the technical debt I have to deal with when working on the project.

Edit: I've had a few ideas regarding this, may end up publishing something in the future.

**Reply by menimeutl** [reply]
if I may ask, in what language are you guys doing this on bc I know C# and want to help out

**Reply by insertish** [reply]
The whole backend is in Rust.

**Comment by phoebehfharris** [comment]
What would this federation be useful for? It seems that most of the similar projects to Revolt are implementing Matrix rather than ActivityPub

**Reply by goliv04053** [reply]
Using a server as a communication server (like Discord or even Matrix) and using a server as a way to access by Revolt ActivityPub and post and interact with people from that network.

**Reply by phoebehfharris** [reply]
Sorry, could you rephrase? How would you be interacting with the rest of the fediverse from Revolt?

**Reply by goliv04053** [reply]
I think that would be interesting to use the server as how I said.. a normal communication server for discord and that you could create an account with a branch of that server. Or about fediverse, make like wordpress or tumblr (use your account as a fediverse connected account)

**Reply by WolfRevo** [reply]
I would break up this topic into two points:
1) federation between different Revolt servers: this would be the main benefit and is similar to federation in Matrix, Mastodon etc., so the federation enables the same service to be interconnected (like my own matrix account on my private matrix server is able to contact any matrix account in the matrix federation, my Mastodon account can read toots from other Mastodon instances my account is not linked with). In context of Revolt my account on my self hosted Revolt server would be easily able to contact users on other Revolt servers out there.
2) one single instance for accounts for different services like Revolt, Matrix, Mastodon etc., but I don't think this would be something important for short term discussions honestly.

**Reply by goliv04053** [reply]
Or maybe a possibility to connect your fediverse account (Mastodon, Calckey and etc.) and access Fediverse on Revolt client.

**Reply by Rexogamer** [reply]
I don't really see how this is feasible - Revolt is a chat platform, while stuff like Mastodon/the -key services are more traditional social media platforms where posts are at least somewhat less frequent/posts are generally longer; the cultures and basic mechanics are pretty different imo

**Reply by nulldg** [reply]
ditto ^

AP is ideal for social media platforms where you've got a profile producing **public** content (eg twitter posts, youtube videos, deviantart images, etc), and you want it to be as discoverable as possible across the fediverse, where it's totally impractical to manually synchronize the things you post across every instance on the internet, and you want to grant custody of your content to a host you trust.
on revolt, there are so many reasons AP would be rather pointless. even if accounts are decentralized, messages would only be located on the instance the server/group is located on. user profiles on revolt contain next to no inherent information, beyond an avatar, name, bio, etc so home servers would be little more than identity servers. this ultimately just adds a pointless middle man.

trying to make revolt to interop with matrix would effectively be forcing a square peg into a round hole - corners would need to be cut off and other things would need to be added (eg encryption) to conform with matrix specs. every platform based on matrix wants their own unique features, and they sometimes end up making themselves partially incompatible with the federation in order to have them, always to the detriment of other clients. often, the bureaucracy moves at snail's pace, like in the case of matrix taking years to add basic things like custom emojis. sometimes progress stalls, and this is what killed IRC.

even without federation, you can always host your own instance and have servers on there. if the revolt team decided to support third-party instances, then revolt would simply need to make it convenient to access multiple instances from the same app in the way IRC did.

**Reply by strypey** [reply]
@nulldg 
> other things would need to be added (eg encryption) to conform with matrix specs

You can connect matrix federation without implementing everything in the spec. Most server clients finish their plaintext implementation before they start adding support for more complex parts of the spec, like encryption. As I noted in my first comment, encryption is in the roadmap for Revolt, and matrix offers a robust, [independently-audited](https://matrix.org/blog/2022/05/16/independent-public-audit-of-vodozemac-a-native-rust-reference-implementation-of-matrix-end-to-end-encryption/) way to do it.

> every platform based on matrix wants their own unique features

This is the precisely the reason for having a common protocol. The devs of each app can implement it in their own way, for the needs of the people using it.

> and they sometimes end up making themselves partially incompatible with the federation in order to have them, always to the detriment of other clients.

Have you got an example of this?

> often, the bureaucracy moves at snail's pace, like in the case of matrix taking years to add basic things like custom emojis. 

This only affects features that require federation between servers, and it doesn't stop you adding features to Revolt, or even federating them between Revolt servers. It just means that users on other apps won't have access to those features, unless their devs copy the way you federate them. If you write up your your implementation as an MSC, others apps can adopt it more easily, whether or not it's accepted and added to the matrix spec.

One more thing, I note that Revolt devs are planning to make their own mobile apps. That's great, but if they implement the client-to-server parts of the matrix spec, people using Revolt would have a wide range of existing matrix mobile apps to choose from. The Revolt crew could even fork one of those and adapt the UI, rather than starting from scratch.

**Comment by strypey** [comment]
I absolutely agree that any client/server communication software ought to be federated. At minimum, with other instances of itself. But if you're going to implement a federation protocol, you might as well use an open protocol, at least for server-to-server federation, respecting people's freedom to choose a different service/ client to chat with people using yours.

I'm a huge fan of ActivityPub. I've been using the fediverse since before it was adopted (back when it was an OStatus network dominated by GNU social servers). But I'm not sure AP is the best fit for federating an instant messaging system like Revolt. Particularly if you're planning on implementing end-to-end encryption, which is not covered at all in the ActivityPub spec. I'd suggest looking at;

* [Matrix](https://spec.matrix.org/): A newer federation protocol, originally developed by the people behind Element. It's now used in Gitter.im and chat.mozilla.org, and RocketChat are in the process of implementing it too, Since Revolt is written in Rust, a [matrix server called Conduit written in Rust](https://gitlab.com/famedly/conduit) could be a good guide. There is also a [matrix-rust-sdk you might be able to use for connecting your clients to matrix servers](https://github.com/matrix-org/matrix-rust-sdk).

* [XMPP](https://xmpp.org/about/technology-overview/) (with OMEMO/MUC): Although less well known, [XMPP is an older protocol standardized by the IETF](https://xmpp.org/rfcs/), based on technology developed for a service called Jabber. Maybe have a look at the [xmpp-rs library for Rust](https://gitlab.com/xmpp-rs/xmpp-rs) as a starting point? XMPP consists of a minimal core spec, supplemented by XEPs that can be implemented on top to support a specific set of features. For example, MUC is an XEP that adds support for group chats, and OMEMO is an XEP that adds supported encryption in both 1:1 chats and group chats. The [ModernXMPP project](https://docs.modernxmpp.org/) documents the implementation of a set of XEPs you need to provide a modern chat experience.

**Reply by goliv04053** [reply]
I think that Matrix is ideal.

**Reply by paulo-roger** [reply]
I am another huge enthusiast of the fediverse, @strypey . Have you seen the chat app the Pixelfed creator is working on? Might bring some ideas for Revolt.

**Reply by paulo-roger** [reply]
After some reading and analising both Matrix and XMPP, I must say XMPP is a lot better than Matrix. XMPP is a solid protocol now, and not in the hands of one corporation.

**Reply by strypey** [reply]
> XMPP is a solid protocol now, and not in the hands of one corporation

Fair comment. The Snikket project shows off the state of the art in XMPP, and its founder Matt - also lead dev of Prosody - is pretty approachable for implementation tips.

The Matrix folks have been working on separating the Matrix Foundation from the VC-funded Element company, and getting more independent server implenters around the table. But either they need to do more on this, or show us more of what's been done.

(EDIT: the [Matrix Manifesto](https://matrix.org/foundation/about/), and this [list of Matrix Foundation members](https://matrix.org/support/), suggests the standard is not really "in the hands of one corporation". Unless you mean the Matrix Foundation itself?)

**Comment by phoebehfharris** [comment]
I'm still not sure that Revolt needs federation - if I want to use Matrix, I can always use Matrix, and if I really need interoperability, a bridge will suffice - but I agree with strypey that federation with a messaging-based protocol would be infinitely more useful than AP.

**Comment by strypey** [comment]
Just giving this a bump. There's been huge growth of interest in decentralised services since 2022. Worth considering some kind of federation with other chat services. Maybe implementing some client-to-server APIs would be an easier place to start than server-to-server federation? It would be great to be able to join Revolt rooms with an XMPP or Matrix app, for example. Even if I still needed a Revolt account to do so.

**Comment by Shanesan** [comment]
Pre-COVID there was little discussion about "federated" anything - today, if your social service is _not_ federated, the people who know anything about Social Media the last couple years give you a quizzical look, like you're about to be bought out at firesale prices by a billionaire at any moment and trodden into the ground.

Having a decentralized server infrastructure and registration system is the future of social media and chat services. I juggle between Matrix and Revolt to self-host (and have installed Matrix to completion) and they have their tasty pieces but both feel like they're missing critical flavors to make a complete dish.

**Comment by strypey** [comment]
The 2.0 version of the Matrix protocol was declared ready to use/late last year;

https://matrix.org/blog/2024/10/29/matrix-2.0-is-here/

Software projects that use Matrix are busy implementing the new features it adds, and now would be a great time to have another look at using Matrix to federate Revolt.
